infer/provider_builder_test.go

```diff
@@ -330,27 +335,32 @@ func TestValidate(t *testing.T) {
 //nolint:paralleltest // Running in parallel causes a data race.
 func TestBuildAndRun(t *testing.T) {
 	// 1. Create a provider without any components and ensure that it returns an error.
-	err := NewProviderBuilder().BuildAndRun()
+	_, err := NewProviderBuilder().Build()
 	require.Error(t, err)
 
 	// 2. Create a provider with a component and ensure that it starts successfully by starting the
 	// provider in a separate goroutine as it blocks the main thread.
 	errChan := make(chan error)
 
 	go func(errCh chan error) {
-		errCh <- NewProviderBuilder().
+		defer close(errCh)
+
+		p, err := NewProviderBuilder().
 			WithComponents(Component(NewMockComponentResource)).
 			WithName("test-provider").
 			WithVersion("1.0.0").
-			BuildAndRun()
+			Build()
+		if err != nil {
+			errCh <- err
+			return
+		}
 
-		close(errCh)
+		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+		defer cancel()
+
+		errCh <- p.Run(ctx)
 	}(errChan)
 
-	select {
-	case err := <-errChan:
-		require.NoError(t, err, "provider startup should not fail")
-	case <-time.After(5 * time.Second):
-		return // The provider started successfully, so we can return.
-	}
+	err = <-errChan
+	assert.NoError(t, err, "provider startup should not fail")
```

Comment:
This is using https://github.com/pulumi/pulumi/pull/19445.

The assertion is stronger because we're now guaranteed the provider executed and return a nil error.